internal/subscriber/ordering_test.go

```diff
@@ -0,0 +1,163 @@
+	"github.com/razorpay/metro/pkg/messagebroker"
+	metrov1 "github.com/razorpay/metro/rpc/proto/v1"
+	"github.com/stretchr/testify/assert"
+)
+
+var mockBatchSize int32 = 1
+var mockOffset int32 = 3
+
+func orderedSetup(t *testing.T, lastStatus string) (
+	ctx context.Context,
+	subImpl *OrderedImplementation,
+) {
+	ctrl := gomock.NewController(t)
+	ctx = context.Background()
+	cs := getMockConsumer(ctx, ctrl)
+	cs.EXPECT().ReceiveMessages(ctx, messagebroker.GetMessagesFromTopicRequest{NumOfMessages: mockBatchSize, TimeoutMs: 1000}).Return(
+		&messagebroker.GetMessagesFromTopicResponse{
+			Messages: getDummyOrderedReceivedMessage(),
+		}, nil,
+	)
+	consumer := getMockConsumerManager(ctx, ctrl, cs)
+	offsetRepo := getMockOffsetRepo(ctrl, lastStatus)
+	subImpl = getMockOrderedImplementation(ctx, consumer, offsetRepo)
+	return
+}
+
+func TestOrderedImplementation_Pull(t *testing.T) {
+	ctx, subImpl := orderedSetup(t, string(sequenceFailure))
+	pullRequest := &PullRequest{ctx: ctx, MaxNumOfMessages: mockBatchSize}
+	responseChan := make(chan *metrov1.PullResponse, 10)
+	errChan := make(chan error, 10)
+	subImpl.Pull(ctx, pullRequest, responseChan, errChan)
+
+	select {
+	case resp := <-responseChan:
+		assert.Zero(t, len(resp.ReceivedMessages))
+		assert.True(t, subImpl.consumer.IsPrimaryPaused(ctx))
+	case <-ticker.C:
+		assert.FailNow(t, "Test case timed out")
+	}
+}
+
+func TestOrderedImplementation_Acknowledge(t *testing.T) {
+	ctx, subImpl := orderedSetup(t, string(sequenceSuccess))
+	pullRequest := &PullRequest{ctx: ctx, MaxNumOfMessages: mockBatchSize}
+	responseChan := make(chan *metrov1.PullResponse, 10)
+	errChan := make(chan error, 10)
+	subImpl.Pull(ctx, pullRequest, responseChan, errChan)
+	select {
+	case resp := <-responseChan:
+		tp := NewTopicPartition(topic, partition)
+		for _, msg := range resp.ReceivedMessages {
+			ackMsg, _ := ParseAckID(msg.AckId)
+			subImpl.Acknowledge(ctx, ackMsg, errChan)
+		}
+		assert.Equal(t, mockOffset, subImpl.consumedMessageStats[tp].maxCommittedOffset)
+	case <-ticker.C:
+		assert.FailNow(t, "Test case timed out")
+	}
+}
+
+func TestOrderedImplementation_ModAckDeadline(t *testing.T) {
+	ctx, subImpl := orderedSetup(t, string(sequenceSuccess))
+	pullRequest := &PullRequest{ctx: ctx, MaxNumOfMessages: mockBatchSize}
+	responseChan := make(chan *metrov1.PullResponse, 10)
+	errChan := make(chan error, 10)
+	subImpl.Pull(ctx, pullRequest, responseChan, errChan)
+
+	select {
+	case resp := <-responseChan:
+		assert.NotEmpty(t, resp.ReceivedMessages)
+		for _, msg := range resp.ReceivedMessages {
+			ackMsg, _ := ParseAckID(msg.AckId)
+			modAckReq := &ModAckMessage{
+				ctx:         ctx,
+				AckMessage:  ackMsg,
+				ackDeadline: int32(0),
+			}
+			subImpl.ModAckDeadline(ctx, modAckReq, errChan)
+		}
+	case <-ticker.C:
+		assert.FailNow(t, "Test case timed out")
+		return
+	}
+	tp := NewTopicPartition(topic, partition)
+	<-time.NewTimer(2 * time.Second).C
```

Comment:
Why is this required? Please avoid wait/sleep on unit tests. CI build times are critical and these add up significantly.